### PR TITLE
Move linear solver from model

### PR DESCRIPTION
**Pull request by Yuriyzabegaev**
## Proposed changes

This moves the linear solver out of the porepy model. The model does not store `linear_system` attribute anymore. `model.assemble_linear_system` now returns it. I have ran the skipped tests.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).
- [x] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: src/porepy/numerics/solvers/nonlinear_solvers.py [hunk 1]
```diff
@@ -70,14 +94,31 @@ class NewtonSolver:
...
 
+        if linear_solver is None:
+            linear_solver = LinearSolverDirect(backend="pypardiso")
+        self.linear_solver: LinearSolverBase = linear_solver
+        """Linear solver object to solve the Jacobian linear systems."""
+        self._linear_solver_initialized = False
+        """Whether model-dependent linear solver state has been initialized.
+        Initialization is done once at :meth:`solve`. It is now assumed that this class
+        must be used with the same model, and should not be reused for a different
+        model.
```
Comment (Yuriyzabegaev) [inline src/porepy/numerics/solvers/nonlinear_solvers.py:121]: This looks ugly and is a temporary solution. We won't need to initialize the linear solver with a model after we have tags and indexers in porepy.

File: src/porepy/numerics/solvers/linear_solver.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    abstract.
+
+    """
+
+    def initialize_with_model(self, model: pp.PorePyModel) -> None:
+        """Initialize model-dependent solver state.
+
+        Solvers without model-dependent state may use this default no-op implementation.
+
+        Note: This is needed in practice to construct a dof manager from a model in
```
Comment (Yuriyzabegaev) [inline src/porepy/numerics/solvers/linear_solver.py:132]: note this comment

File: src/porepy/numerics/solvers/nonlinear_solvers.py [hunk 3]
```diff
@@ -391,12 +467,26 @@ def iteration(self, model: SolutionStrategy) -> np.ndarray:
...
-        nonlinear_increment = model.solve_linear_system()
-        return nonlinear_increment
+        linear_system = model.assemble_linear_system()
+        nonlinear_increment, linear_solver_status = (
+            self.linear_solver.solve_linear_system(linear_system)
+        )
+
+        # This is a temporary approach for compatability. It is a linear solver's
+        # responsibility and will be moved to a new linear solver class when tags and
+        # indexer are introduced.
```
Comment (Yuriyzabegaev) [inline src/porepy/numerics/solvers/nonlinear_solvers.py:480]: note this comment

File: src/porepy/viz/data_saving_model_mixin.py [hunk 4]
```diff
@@ -345,7 +345,7 @@ def after_nonlinear_iteration(self, solution_vector: np.ndarray) -> None:
         iterate subdictionary.
 
         """
-        super().after_nonlinear_iteration(solution_vector)  # type: ignore[misc]
+        super().after_nonlinear_iteration(solution_vector)  # type: ignore[safe-super]
```
Comment (Yuriyzabegaev) [inline src/porepy/viz/data_saving_model_mixin.py:348]: specified what exactly we ignore because suddenly mypy started disliking how it was before. Probably because I added `after_nonlinear_iteration` in a protocol.

File: tests/models/test_solution_strategy.py [hunk 5]
```diff
@@ -642,3 +652,18 @@ def set_domain(self) -> None:
         )
         and np.all(approx_identity.shape == identity.shape)
     )
+
+    # Test that we expand the solution correctly. This is currently hard-coded in
```
Comment (Yuriyzabegaev) [inline tests/models/test_solution_strategy.py:660]: I extended this test to make sure we don't break the full pipeline of solving the model with Schur complement reduction. It made this test 1.5 times slower, but this test will be reconsidered when we change where the Schur complement is constructed.

File: tests/numerics/solvers/test_linear_solver.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        return
+
+    mat = linear_system.matrix
+    rhs = linear_system.rhs
+    assert mat is not None
+    if case == "nonsingular":
+        assert status.is_success()
+        assert np.allclose(mat @ sol - rhs, 0, rtol=0, atol=1e-10)
+    elif case == "singular":
+        # Returned array can have anything inside, but with the right shape and dtype.
```
Comment (Yuriyzabegaev) [inline tests/numerics/solvers/test_linear_solver.py:68]: pypardiso does not return an error or nans if the problem is singular, it returns finite values of large magnitude. Therefore, the linear solver has no way to tell that it failed and returns success. Otherwise, we'd have to introduce the logic of divergence tolerances on the linear solver api level, and I really don't see a reason for it.

This is fine, as these values will be discarded by a divergence check on the nonlinear level.

File: tutorials/solution_strategies.ipynb [hunk 7]
```diff
@@ -62,7 +62,7 @@
             "    \"meshing_arguments\": {\"cell_size\": 1 / 4},\n",
             "}\n",
             "solver_params = {\n",
-            "    \"nl_convergence_tol_res\": 1e-10,  # Since the line search reduces the update, a residual based tolerance is needed\n",
+            "    \"nl_convergence_res_atol\": 1e-10,  # Since the line search reduces the update, a residual based tolerance is needed\n",
```
Comment (Yuriyzabegaev) [inline tutorials/solution_strategies.ipynb:64]: Just updated the key

File: src/porepy/numerics/solvers/linear_solver.py (hunk 2 above)
Comment (keileg) [reply src/porepy/numerics/solvers/linear_solver.py:132]: Let's discuss this, just to make sure I understand the context.

File: src/porepy/numerics/solvers/nonlinear_solvers.py [hunk 8]
```diff
@@ -267,16 +324,25 @@ def solve(self, model: SolutionStrategy) -> NonlinearSolverStatus:
             The status of the nonlinear solver.
 
         """
+        # Model-dependent setup of a linear solver is done once.
```
Comment (keileg) [inline src/porepy/numerics/solvers/nonlinear_solvers.py:330]: Let's talk through the scenarios here. I'm not fully convinced we do not want to alter the (type of) linear solver in between Newton iterations.

File: src/porepy/numerics/solvers/nonlinear_solvers.py (hunk 3 above)
Comment (keileg) [reply src/porepy/numerics/solvers/nonlinear_solvers.py:480]: Agreed.

File: src/porepy/__init__.py [hunk 9]
```diff
@@ -54,6 +54,14 @@
 )
 from porepy.utils import array_operations
 from porepy.numerics.linalg import matrix_operations
+from porepy.numerics.linalg.linear_solver import (
```
Comment (keileg) [inline src/porepy/__init__.py]: This worries me somewhat. I see three alternatives:
1. Accept further cluttering of the top-level namespace.
2. 1., but try to limit the classes exposed at the top level.
3. Introduce a second level, similar to pp.ad, and use this for all functionality for solvers/numerics (not including existing discretizations). The name for this second level could, unfortunately, be pp.solvers. 
Let's discuss.

File: tests/numerics/nonlinear/test_nonlinear_solvers.py [hunk 10]
```diff
@@ -62,15 +68,17 @@ def time_step_status_in_progress() -> TimeStepperStatusContinueIterating:
...
             convergence_statuses=ConvergenceStatusCollection(),
             divergence_statuses=ConvergenceStatusCollection(),
         ),
     )
 
 
-@pytest.fixture
-def default_newton_solver():
+def default_newton_solver(nonlinear_increment_history: Optional[np.ndarray] = None):
+    # TODO
```
Comment (keileg) [inline tests/numerics/nonlinear/test_nonlinear_solvers.py]: Please expand or delete.

File: tests/numerics/solvers/test_nonlinear_solvers.py [hunk 11]
```diff
@@ -62,15 +68,17 @@ def time_step_status_in_progress() -> TimeStepperStatusContinueIterating:
...
+            linear_solver_statuses=linear_solver_statuses(2),
             convergence_statuses=ConvergenceStatusCollection(),
             divergence_statuses=ConvergenceStatusCollection(),
         ),
     )
 
 
-@pytest.fixture
-def default_newton_solver():
+def default_newton_solver(nonlinear_increment_history: Optional[np.ndarray] = None):
```
Comment (keileg) [inline tests/numerics/solvers/test_nonlinear_solvers.py:78]: Any reason this is no longer a fixture? Statistics?

File: tests/utils/test_ui_and_logging.py [hunk 12]
```diff
@@ -82,25 +82,33 @@ def before_nonlinear_iteration(self) -> None:
...
+        return pp.LinearSystem(csr_matrix((0, 0)), np.ones(shape=()))
 
-    def solve_linear_system(self) -> np.ndarray:
-        # Artificially fail/satisfy Newton update norm convergence criterion.
-        # IMPLEMENTATION NOTE nonlinear_solver_statistics.num_iterations lags behind
-        # one iteration at this point in the solver step.
-        if self.nonlinear_solver_statistics.num_iterations < self.num_nl_iterations - 1:
-            return np.array([1.0])
+
+class MockLinearSolver(pp.LinearSolverBase):
```
Comment (keileg) [inline tests/utils/test_ui_and_logging.py]: Would it be meaningful to gather these mock objects used for tests in some common utility file?

File: tests/utils/test_ui_and_logging.py [hunk 13]
```diff
@@ -138,11 +146,18 @@ def run_model_and_save_output(
...
+    model = MockModel(num_time_steps)
     params = {"progressbars": progressbars}
+    nonlinear_solver = pp.NewtonSolver(
+        params=params, linear_solver=MockLinearSolver(num_nl_iterations)
+    )
 
     # MockModel has the wrong type. Ignore mypy.
-    pp.ModelRunner(model, params).run()  # type: ignore
+    pp.ModelRunner(
+        model,  # type: ignore
```
Comment (keileg) [inline tests/utils/test_ui_and_logging.py]: No need for this, we will not run mypy on the tests (the cost-benefit analysis is in all likelyhood negative for this).

File: tests/numerics/solvers/test_linear_solver.py (hunk 6 above)
Comment (keileg) [reply tests/numerics/solvers/test_linear_solver.py:68]: Just to be sure: I see that pypardiso can return error codes. This is not something we want to exploit at this stage, right? I realize the question is wrongly placed.

File: tests/viz/test_solver_statistics.py [hunk 14]
```diff
@@ -58,16 +58,19 @@ def nonlinear_solver_status(
 ) -> NonlinearSolverStatus:
     """Create a nonlinear solver status for the tests."""
     if status == "converged":
-        status_type = NonlinearSolverStatusConverged
+        return NonlinearSolverStatusConverged(
+            linear_solver_statuses=[pp.LinearSolverStatusSuccess(solve_time=0.0)] * 2,
```
Comment (keileg) [inline tests/viz/test_solver_statistics.py]: Randomly placed comment: I realize more and more I am not in control of the may we treat statistics or various kinds. Should we have a deep discussion of this now, or is it better left at a later point, towards the end of the solver redesign?

File: src/porepy/numerics/solvers/nonlinear_solvers.py (hunk 1 above)
Comment (keileg) [reply src/porepy/numerics/solvers/nonlinear_solvers.py:121]: Fine for now. Let's just not forget it.

**Review by keileg — APPROVED** [review]
Approved, but let's discuss various comments briefly before you merge.
